### PR TITLE
Fix client-side race in `docker stats`

### PR DESCRIPTION
**Pull request by icecrime**
Subscribe to events and monitor for new containers before the initial
listing of currently running containers.

This fixes a race where a new container could appear between the first
list call but before the client was subscribed to events, leading to a
container never appearing in the output of `docker stats`.

### CODE REVIEW

**Comment by icecrime** [comment]
This should fix the CI failures on `TestStatsAllNewContainersAdded`, particularly common on win2lin because of the extra latency between client (Azure) and daemon (DO).

File: api/client/stats.go [hunk 1]
```diff
@@ -249,9 +228,37 @@ func (cli *DockerCli) CmdStats(args ...string) error {
...
 		}(closeChan)
+
+		// Block until we're subscribed to the events endpoint.
+		<-started
 	} else {
 		close(closeChan)
 	}
+
+	if showAll {
+		options := types.ContainerListOptions{
```
Comment (LK4D4) [inline api/client/stats.go]: Maybe for clarity it's better to move this code block right after `<-started`

**Comment by LK4D4** [comment]
That func is a beast.

**Comment by icecrime** [comment]
The fix is not working... Looking into it.

**Comment by icecrime** [comment]
Updated: it's still ugly as hell, but the concurrency model is much easier to reason about: a single channel centralizes all requests to add/remove containers from the `stats` output.

- When invoked without arguments: a long-running goroutine is started to monitor all container events (pushing creation/removal messages to that central channel), and a short-lived one is fired to emit creation messages for the list of previously existing containers.
- When invoked for a set of containers: we emit creation messages for each requested container, and wait for a bit to detect errors (as was previously done).

**Comment by jfrazelle** [comment]
LGTM

File: api/client/stats.go [hunk 2]
```diff
@@ -150,123 +158,151 @@ func (cli *DockerCli) CmdStats(args ...string) error {
 	names := cmd.Args()
 	showAll := len(names) == 0
 
-	if showAll {
+	// The containerChan is the central synchronization piece for this function,
```
Comment (jfrazelle) [inline api/client/stats.go]: This is a v pretty comment

**Comment by runcom** [comment]
LGTM

**Comment by estesp** [comment]
I assume PRs created today (2/29) are using this fix, but sadly I just hit the error in https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/22473/console

Not sure it helps as there is no real output as to what went wrong..

**Comment by icecrime** [comment]
@estesp :cry:

**Pull request merged**